lua/starfall/toolscreen.lua

```diff
@@ -0,0 +1,211 @@
+local star_reset_radius      = 200                                -- Distance to which teleport the star when flipping
+local star_reset_radius_sqr  = star_reset_radius ^ 2 + 10         -- Distance squared and offset at which to teleport the star
+local star_random_offset     = math.pi / 8                        -- Random angle offset to set when flipping the star
+local function get_random_star_properties()
+	return
+		HSVToColor(math_rand(190, 220), math_rand(0.6, 1), 0.9),  -- Color
+		math_rand(40, 120),                                       -- Size
+		math_rand(200, 600),                                      -- Gravity
+		math_rand(0.3, 1) * (math_random() > 0.5 and 1 or -1)     -- Rotation direction
+end
+
+-------------------------------------------
+
+local stars = {}
+for _ = 1, star_count do
+	local color, size, gravity, ang_dir = get_random_star_properties()
+	stars[#stars+1] = {
+		x       = math_rand(-128, 128),
+		y       = math_rand(-128, 128),
+		x_vel   = math_rand(-star_velocity_max, star_velocity_max),
+		y_vel   = star_velocity_max,
+		ang     = 0,
+		ang_vel = 0,
+		ang_dir = ang_dir,
+		color   = color,
+		size    = size,
+		gravity = gravity,
+	}
+end
+
+local overlay_material = Material("radon/starfall_tool_overlay.png", "smooth")
+overlay_material:SetInt("$flags", 0x8000 + 0x0020)
+
+local star_material = Material("radon/starfall_tool_star.png", "smooth")
+star_material:SetInt("$flags", 0x8000 + 0x0080 + 0x0020 + 0x0010)
+
+-- RenderTargets created before rendering is ready suffer from depth related issues
+local star_canvas, star_canvas_material
+hook.Add("PreRender", "StarfallToolscreenPrepare", function()
+	star_canvas = GetRenderTarget("starfall_tool_canvas", 256, 256)
+	star_canvas_material = CreateMaterial("starfall_tool_material", "UnlitGeneric", { ["$basetexture"] = star_canvas:GetName() })
+	star_canvas_material:SetInt("$flags", 0x8000 + 0x0020 + 0x0010)
+	star_canvas_material:SetInt("$ignorez", 1)
+	hook.Remove("PreRender", "StarfallToolscreenPrepare")
+end)
+
+local stars_errored_delay = 0
+hook.Add("StarfallError", "StarfallToolscreenError", function(_, owner, client)
+	local local_player = LocalPlayer()
+	if owner ~= local_player then return end
+	if client and client ~= local_player then return end
+
+	if stars_errored_delay <= 0 then
+		for _, star in ipairs(stars) do
+			star.color = HSVToColor(350 + math_random() * 20, 0.7, 0.85)
+		end
+	end
+	stars_errored_delay = 3 -- How long should the stars remain red after an error
+end)
+
+-------------------------------------------
+
+local last_frame = 0
+local is_linux = system.IsLinux()
+local ply_eye_ang_prev = Angle()
+function SF.DrawToolgunScreen(w, h, title, scroll_text)
+	local curtime = RealTime()
+	local dt = curtime - last_frame
+	if dt > 1 / simulation_fps then
+		last_frame = curtime
+		stars_errored_delay = stars_errored_delay - dt
+
+		local deceleration = star_deceleration ^ dt
```

Comment:
Is this supposed to be multiplied?

Reply:
No, that's a correct way to approximate the deceleration based with deltatime. That's why I localized, kinda costly.